src/Faithlife.Utility/WrappingStreamBase.cs

```diff
@@ -0,0 +1,166 @@
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Faithlife.Utility
+{
+	/// <summary>
+	/// A <see cref="Stream"/> that wraps another stream. One major feature of this class is that it does not dispose the underlying stream
+	/// when it is disposed if <see cref="Ownership.None"/> is used; this is useful when using classes such as <see cref="BinaryReader"/>
+	/// that take ownership of the stream passed to their constructors. This class delegates a minimal number of properties and methods to
+	/// the wrapped stream so that inheritors can override this class as they would <see cref="Stream"/>, e.g. override Read and get
+	/// correct behavior for ReadAsync and CopyToAsync.
```

Comment:
I'm not sure that replacing asynchronous I/O with synchronous I/O should be considered "correct behavior", which (I believe) is what this change does by not forwarding the async methods to the wrapped stream.

Reply:
The inheriting rules are the same as for `Stream`. You must override `Read` (if you want to change its behavior). You may override `ReadAsync`, etc. if you want to optimize those paths, otherwise you get the default implementations, which work but may not be optimal.

Reply:
https://docs.microsoft.com/en-us/dotnet/api/system.io.stream?view=netframework-4.7.1#Notes_to_Inheritors

Reply:
I see what you mean, but breaking asynchronous I/O just by inheriting the class still doesn't seem like a good plan to me.

I wonder if we should `abstract` any member that the derived class probably has to implement to properly wrap both sync and async I/O. Methods that are probably fine with their default `Stream` implementation can be left alone. No method delegates to the wrapped stream by default; the derived class has to decide.

Reply:
Well, maybe we could pass through `CanTimeout`, `Flush()`, `FlushAsync()`, `ReadTimeout`, and `WriteTimeout`, since that's almost certainly the right behavior, and almost certainly safe.

Reply:
We could maybe seal `Close()`, since it shouldn't be overridden. Also, we're probably best off sealing the old-school `(Begin/End)(Read/Write)`, either using the default `Stream` implementation (which uses synchronous I/O) or coming up with an implementation that uses `(Read|Write)Async` (probably not worth it).

Reply:
So we leave `CopyToAsync`, `ReadByte`, and `WriteByte` alone -- derived class can override if they want, but default implementation should be fine.

Everything else is `abstract`: `Can(Read|Write|Seek)`, `(Set)?Length`, `Position|Seek`, `(Read|Write)(Async)?`.

Reply:
If we're in agreement, I'm happy to make the changes; let me know. @ddunkin @bgrainger

Reply:
I understand making Read/WriteAsync abstract to prevent accidental deoptimization, but not the purpose of making CanRead, etc. abstract, unless it's to make inheritors be very explicit.

Reply:
It would be to force them to consider all of the important overridables, yes, but perhaps it is overkill, and only `(Read|Write)(Async)?` need to be abstract, and the others I mentioned can just pass through.

Reply:
(I can imagine a derived class implementer forgetting that the caller might seek the stream or set the length.)